pkg/virt-api/rest/subresource.go

```diff
@@ -908,48 +908,26 @@ func (app *SubresourceAPIApp) FilesystemList(request *restful.Request, response
 }
 
 func generateVMVolumeRequestPatch(vm *v1.VirtualMachine, volumeRequest *v1.VirtualMachineVolumeRequest) (string, error) {
-	verb := "add"
-	if len(vm.Status.VolumeRequests) > 0 {
-		verb = "replace"
-	}
-
+	verb := getPatchVerb(vm.Status.VolumeRequests)
 	vmCopy := vm.DeepCopy()
 
 	// We only validate the list against other items in the list at this point.
 	// The VM validation webhook will validate the list against the VMI spec
 	// during the Patch command
 	if volumeRequest.AddVolumeOptions != nil {
-		name := volumeRequest.AddVolumeOptions.Name
-		for _, request := range vm.Status.VolumeRequests {
-			if request.AddVolumeOptions != nil && request.AddVolumeOptions.Name == name {
-				return "", fmt.Errorf("Add volume request for volume [%s] already exists", name)
-			} else if request.RemoveVolumeOptions != nil && request.RemoveVolumeOptions.Name == name {
-				return "", fmt.Errorf("Unable to add volume. A remove volume request for volume [%s] already exists and is still being processed.", name)
-			}
+		if err := addAddVolumeRequests(vm, volumeRequest, vmCopy); err != nil {
+			return "", err
 		}
-		vmCopy.Status.VolumeRequests = append(vm.Status.VolumeRequests, *volumeRequest)
 	} else if volumeRequest.RemoveVolumeOptions != nil {
-		name := volumeRequest.RemoveVolumeOptions.Name
-		volumeRequestsList := []v1.VirtualMachineVolumeRequest{}
-		for _, request := range vm.Status.VolumeRequests {
-			if request.AddVolumeOptions != nil && request.AddVolumeOptions.Name == name {
-				// Filter matching AddVolume requests from the new list.
-				continue
-			} else if request.RemoveVolumeOptions != nil && request.RemoveVolumeOptions.Name == name {
-				return "", fmt.Errorf("A remove volume request for volume [%s] already exists and is still being processed.", name)
-			}
-
-			volumeRequestsList = append(volumeRequestsList, request)
+		if err := addRemoveVolumeRequests(vm, volumeRequest, vmCopy); err != nil {
+			return "", err
```

Comment:
Here, we could have two functions:
1 )  a predicate function `shouldAddVolumeRequests` for  request.AddVolumeOptions != nil && request.AddVolumeOptions.Name == name
2) validateRemoveVolumeRequest() 

And then again, leave the for loop here and just check if we should add request, validate and finally `volumeRequestsList = append(volumeRequestsList, request)
`

What do you think?

Reply:
I agree with 1 )
Regarding 2 ), I think `validateRemoveVolumeRequest()` would make it overly complicated as in each iteration of `for`,   there is another `append` to an array. Imo it is more readable when the changes made to the array are not "hidden" inside another function.

Reply:
Ack